### PR TITLE
overlord/configstate: change how ssh is stopped/started (2.32)

### PR DESCRIPTION
**Pull request by zyga**
The ssh.service has an alias of sshd.service and so we must operate on
the main file ssh.service for enable/disable operations as aliases do
not exist when the main unit is disabled.

Due to some unfortunate use of writable paths we cannot mask ssh.service
and have that unit stay gone on reboot. To work around that use the fact
that the unit is conditional of non-existence of /etc/ssh/sshd_not_to_be_run

The patch also unmasks sshd.service in case it was masked by earlier
snapd.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: overlord/configstate/configcore/services.go [hunk 1]
```diff
@@ -37,22 +40,36 @@ func (l *sysdLogger) Notify(status string) {
...
+	sshCanary := filepath.Join(dirs.GlobalRootDir, "/etc/ssh/sshd_not_to_be_run")
 
 	switch value {
 	case "true":
-		if err := sysd.Disable(serviceName); err != nil {
-			return err
-		}
-		if err := sysd.Mask(serviceName); err != nil {
-			return err
+		if serviceName == "ssh.service" {
```
Comment (niemeyer) [inline overlord/configstate/configcore/services.go]: Thanks for finding a workaround on short notice, Zygmunt.

I heard you've made this work in practice, which is great, but I'm still a bit lost here. It's not clear to me what this is doing yet, and how it makes sense to error saying that snapd configuration has it a given way when this is in fact the snapd configuration.

Let's please discuss this tomorrow.

File: overlord/configstate/configcore/services.go [hunk 2]
```diff
@@ -37,22 +40,36 @@ func (l *sysdLogger) Notify(status string) {
...
 		}
 		return sysd.Stop(serviceName, 5*time.Minute)
 	case "false":
-		if err := sysd.Unmask(serviceName); err != nil {
-			return err
-		}
-		if err := sysd.Enable(serviceName); err != nil {
-			return err
+		if serviceName == "ssh.service" {
+			err := os.Remove(sshCanary)
```
[the remark this answers is not in the document]
Comment (zyga) [reply overlord/configstate/configcore/services.go]: This is done now.

**Comment by zyga** [comment]
FYI, @bboozzoo pointed this out, https://github.com/snapcore/core/commit/dc782a37a3aef3fb11f256260f5205f2ff256acd -- we used to do this before we had in-core config mechanism.

File: overlord/configstate/configcore/services.go [hunk 3]
```diff
@@ -33,9 +36,37 @@ func (l *sysdLogger) Notify(status string) {
...
+		return sysd.Start(serviceName)
+	default:
+		return fmt.Errorf("option %q has invalid value %q", serviceName, value)
+	}
+}
+
+// switchDisableTypicalService switches a service in/out of disabled state
 // where "true" means disabled and "false" means enabled.
-func switchDisableService(serviceName, value string) error {
+func switchDisableTypicalService(serviceName, value string) error {
```
Comment (niemeyer) [inline overlord/configstate/configcore/services.go]: No need to split that one out. We can just have this as the real switchDisableService, and do the trivial if as the first line of it.

**Pull request merged**